File: examples/vbnet/ex_leader.vb
Imports Rhino
Imports Rhino.Geometry
Imports Rhino.Commands
Imports System.Collections.Generic
Imports System.Linq

Namespace examples_vb
  <System.Runtime.InteropServices.Guid("055686DA-E4DF-4241-99AF-212546C26F08")> _
  Public Class LeaderCommand
    Inherits Command
    Public Overrides ReadOnly Property EnglishName() As String
      Get
        Return "vbLeader"
      End Get
    End Property

    Protected Overrides Function RunCommand(doc As RhinoDoc, mode As RunMode) As Result
      Dim points = New List(Of Point3d)() From { _
        New Point3d(1, 1, 0), _
        New Point3d(5, 1, 0), _
        New Point3d(5, 5, 0), _
        New Point3d(9, 5, 0) _
      }

      Dim xyPlane = Plane.WorldXY

      Dim points2d = New List(Of Point2d)()
      For Each point3d As Point3d In points
        Dim x As Double, y As Double
        If xyPlane.ClosestParameter(point3d, x, y) Then
          Dim point2d = New Point2d(x, y)
          If points2d.Count < 1 OrElse point2d.DistanceTo(points2d.Last()) > RhinoMath.SqrtEpsilon Then
            points2d.Add(point2d)
          End If
        End If
      Next

      doc.Objects.AddLeader(xyPlane, points2d)
      doc.Views.Redraw()
      Return Result.Success
    End Function
  End Class
End Namespace

File: examples/cs/ex_leader.cs
using Rhino;
using Rhino.Geometry;
using Rhino.Commands;
using Rhino.Input.Custom;
using System.Collections.Generic;
using System.Linq;

namespace examples_cs
{
  [System.Runtime.InteropServices.Guid("891F5AE0-DBE0-40A2-8C28-F59B8A757933")]
  public class ex_leader : Rhino.Commands.Command
  {
    public override string EnglishName { get { return "csLeader"; } }

    protected override Rhino.Commands.Result RunCommand(RhinoDoc doc, Rhino.Commands.RunMode mode)
    {
      var points = new List<Point3d>();
      points.Add(new Point3d(1, 1, 0));
      points.Add(new Point3d(5, 1, 0));
      points.Add(new Point3d(5, 5, 0));
      points.Add(new Point3d(9, 5, 0));

      var xyPlane = Plane.WorldXY;

      var pts2d = new List<Point2d>();
      foreach (var pt3d in points)
      {
        double x, y;
        if (xyPlane.ClosestParameter(pt3d, out x, out y))
        {
          var pt2d = new Point2d(x, y);
          if (pts2d.Count < 1 || pt2d.DistanceTo(pts2d.Last<Point2d>()) > RhinoMath.SqrtEpsilon)
            pts2d.Add(pt2d);
        }
      }

      doc.Objects.AddLeader(xyPlane, pts2d);
      doc.Views.Redraw();
      return Rhino.Commands.Result.Success;
    }
  }
}